walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,66 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
 
+/** Finds potential meeting times. */
 public final class FindMeetingQuery {
+
+  /**
+   * Returns available TimeRanges for the {@code request} given {@code events}, a Collection of
+   * all events so far. 
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> times = new ArrayList<>();
+    Collection<String> attendees = request.getAttendees();
+    
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      // If the event is longer than a day, no time slots are available to book a meeting.
+      return times; 
+    } else if (attendees.size() == 0) {
+      // If there are no attendees, the entire day is available to book a meeting.
+      times.add(TimeRange.WHOLE_DAY);
+      return times;
+    }
+    int currentEndTime = TimeRange.START_OF_DAY;
+    for (Event event : events) {
+      // Ignores this event's time range if the people attending this event are 
+      // not attending the requested meeting. 
+      Set<String> eventAttendees = new HashSet<String>(event.getAttendees());
+      eventAttendees.retainAll(attendees);
+      if (eventAttendees.isEmpty()) {
+        continue;
+      }
+
+      TimeRange time = event.getWhen();
+      // Skips this event because it ended before or at the same time as
+      // the latest event.
+      if (time.end() <= currentEndTime) {
+        continue;
+      } else if (time.start() <= currentEndTime) {
+        // Handles events starting before but ending after the currentEndTime.
+        currentEndTime = time.end();
+        continue;
+      }else if (request.getDuration() > (time.start() - currentEndTime)) {
```

Comment:
This statement can be merged with the previous one because the same outcome occurs

Reply:
Done!